pkg/auth/providers/azure/azure_actions.go

```diff
@@ -118,28 +118,56 @@ func (ap *azureProvider) testAndApply(actionName string, action *types.Action, r
 	return ap.tokenMGR.CreateTokenAndSetCookie(user.Name, userPrincipal, groupPrincipals, providerToken, 0, "Token via Azure Configuration", request, userExtraInfo)
 }
 
-// migrateToMicrosoftGraph represents the migration of the registered Azure AD auth provider
+// migrateToMicrosoftGraph performs the migration of the registered Azure AD auth provider
 // from the deprecated Azure AD Graph API to the Microsoft Graph API.
 // It modifies the existing auth config value in the database, so that it has up-to-date endpoints to the new API.
 // Most importantly, it sets the annotation that specifies that the auth config has been migrated to use the new auth flow.
+
+// It also verifies that admins have properly configured an existing app registration's permissions
+// in the Azure portal before they update their Azure AD auth config to use the new authentication flow.
+// The method receives the current AuthConfig from the database, then updates it in-memory to use the new endpoints,
+// and creates a new test Azure client, thereby getting an access token to the Graph API.
+// Then it parses the JWT and inspects the permissions contained within. If the admins had not set those as per docs,
+// then Rancher won't find the permissions in the test token and will return an error.
+
+// If validation and applying work, then migrateToMicrosoftGraph deletes all secrets with access tokens to the
+// deprecated Azure AD Graph API.
 func (ap *azureProvider) migrateToMicrosoftGraph() error {
-	defer ap.deleteUserAccessTokens()
-	defer clients.GroupCache.Purge()
+	cfg, err := ap.updateConfigAndValidatePermissions()
+	if err != nil {
+		return err
+	}
+	if err = ap.applyUpdatedConfig(cfg); err != nil {
+		return err
+	}
+	ap.deleteUserAccessTokens()
+	clients.GroupCache.Purge()
+	return nil
+}
 
+func (ap *azureProvider) updateConfigAndValidatePermissions() (*v32.AzureADConfig, error) {
 	cfg, err := ap.getAzureConfigK8s()
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if authProviderEnabled(cfg) {
-		updateAzureADEndpoints(cfg)
+	updateAzureADEndpoints(cfg)
+
+	// Try to get a new client, which will fetch a new access token and validate its permissions.
+	_, err = clients.NewMSGraphClient(cfg, ap.secrets)
```

Comment:
This seems a bit risky to me. Before, I think the client was only constructed when we had an active, enabled config. But here we are constructing even if the config is not active. This seems like it could easily result in a nil pointer due to some value not being filled in that we expect to be filled in. Do you consider this function safe to call for a non-enabled auth config?

Reply:
I'd agree. The `authProviderEnabled` check got moved to later in the logic, and it should still be up front.

Reply:
That's a good point! I checked and had no panics. The upgrade request gets an error on trying to initialize the client, since the secret exchange fails (the secret is an empty string in a disabled config). But I added a check if the auth provider is enabled before trying to initialize a client to the Graph API.

The only small thing is the response code, as I'd prefer to have a 400 in this case, but returning an error from this area of the code always gets caught in a generic 500 response with code. I am not sure how consistent we are in these situations elsewhere...

Reply:
I'm guessing that if you directly return an httperror it will take the code from that error rather than wrapping in a 500 (see the `ActionNotAvailable` error in `actionHandler` and the `InvalidBodyCount` error in `testAndApply`). Would be worth giving it a try if you would prefer a specific 400 error code.

Reply:
I think `ActionNotAvailable` is good here, but its code is 404, which might seem strange, but better than 500.
There is also `InvalidState 422`, which could be reasonable here.